### PR TITLE
[SPARK-6055] [PySpark] fix incorrect __eq__ of DataType

### PR DESCRIPTION
**Pull request by davies**
The _eq_ of DataType is not correct, class cache is not use correctly (created class can not be find by dataType), then it will create lots of classes (saved in _cached_cls), never released.

Also, all same DataType have same hash code, there will be many object in a dict with the same hash code, end with hash attach, it's very slow to access this dict (depends on the implementation of CPython).

This PR also improve the performance of inferSchema (avoid the unnecessary converter of object). 

cc @pwendell  @JoshRosen

### CODE REVIEW

**Comment by davies** [comment]
will create another PR for 1.2 and 1.1

File: python/pyspark/sql/context.py [hunk 1]
```diff
@@ -620,93 +619,6 @@ def _get_hive_ctx(self):
         return self._jvm.HiveContext(self._jsc.sc())
 
 
-def _create_row(fields, values):
```
Comment (davies) [inline python/pyspark/sql/context.py]: These are duplicated, also in types.py.

Comment (JoshRosen) [inline python/pyspark/sql/context.py]: Yep, good catch.

File: python/pyspark/sql/types.py [hunk 2]
```diff
@@ -242,11 +240,12 @@ def __init__(self, elementType, containsNull=True):
         :param elementType: the data type of elements.
         :param containsNull: indicates whether the list contains None values.
 
-        >>> ArrayType(StringType) == ArrayType(StringType, True)
+        >>> ArrayType(StringType()) == ArrayType(StringType(), True)
```
Comment (JoshRosen) [inline python/pyspark/sql/types.py]: Is this a breaking API change?  Or were the old doctests showing incorrect usage of the API?

Comment (davies) [inline python/pyspark/sql/types.py]: Old tests are incorrect.

File: python/pyspark/sql/types.py [hunk 3]
```diff
@@ -505,6 +508,9 @@ def __eq__(self, other):
 
 def _parse_datatype_json_string(json_string):
     """Parses the given data type JSON string.
+    >>> import pickle
```
Comment (JoshRosen) [inline python/pyspark/sql/types.py]: It looks like this is a regression test for the singleton `__eq__` issue.  It feels a little odd to have this as a doctest as opposed to a dedicated `unittest` regression test.  Maybe we could move this into a new test case and add a comment referencing this JIRA?

Comment (JoshRosen) [inline python/pyspark/sql/types.py]: e.g. maybe put it in `sql/tests.py`?

Comment (davies) [inline python/pyspark/sql/types.py]: make sense, will do.

File: python/pyspark/sql/types.py [hunk 4]
```diff
@@ -786,8 +792,24 @@ def _merge_type(a, b):
...
+        return _need_converter(dataType.keyType) or _need_converter(dataType.valueType)
+    elif isinstance(dataType, NullType):
+        return True
+    else:
+        return False
+
+
 def _create_converter(dataType):
     """Create an converter to drop the names of fields in obj """
+    if not _need_converter(dataType):
```
Comment (JoshRosen) [inline python/pyspark/sql/types.py]: Is this particular call to `_need_converter()` acting as an optimization?  It performs the same sequence of `isinstance` checks as the code below, so don't know if it's actually saving any work in this case.

File: python/pyspark/sql/types.py [hunk 5]
```diff
@@ -64,6 +64,8 @@ def json(self):
                           sort_keys=True)
 
 
+# This singleton pattern does not work with pickle, you will get
```
Comment (JoshRosen) [inline python/pyspark/sql/types.py]: Would it be possible to have singletons that pickled properly if we implemented a custom `__reduce__()` method?  Was the old `__eq__` behavior an important performance optimization?

**Comment by JoshRosen** [comment]
It looks like `_restore_object` still tries to use DataType instance `id`s as `_cached_cls` dictionary keys during unpickling; is this still necessary if the DataTypes aren't singletons after unpickling?

File: python/pyspark/sql/types.py (hunk 4 above)
Comment (davies) [inline python/pyspark/sql/types.py]: It's useful for ArrayType and MapType, can it didnot can `lambda x:x` on all the objects in it.

Comment (JoshRosen) [inline python/pyspark/sql/types.py]: Ah, right; I overlooked that.  Makes sense.

File: python/pyspark/sql/types.py (hunk 5 above)
Comment (davies) [inline python/pyspark/sql/types.py]: We discussed this, the old implementation is a bit of over optimization, I think.

For PrimitiveTypeSingleton, the __dict__ is empty, so the new implementation is not so heavy, so there will be no much performance difference.

**Comment by davies** [comment]
@JoshRosen Because we serialized the objects in batch, and pickle memorize the multiple occurrences of same object  in the batch, finally we will get single DataType object (even for StructType), we can benefits from this optimization, no `__hash__` and `__eq__` for later row in the batch.

**Comment by JoshRosen** [comment]
Makes sense; LGTM.  I'll take a look at the backport patches, too.

**Comment by JoshRosen** [comment]
LGTM, so I've merged this into `branch-1.3` (1.3.0) and `master` (1.4.0).  Thanks!

**Pull request closed without merging**